### PR TITLE
nixifying the cdk node

### LINKED ISSUE #1618: Nix file that allows to install / configure all the dependencies

### PR DESCRIPTION
**Pull request by akonring**
Closes EspressoSystems/espresso-sequencer#1618

### This PR:

Introduces nix config: `flake.nix`, `flake.lock` and `shell.nix` to be consistent with other espresso repositories.
### This PR does not:

Do anything else

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+              http2Support = false;
+              libssh2 = muslPkgs.pkgsStatic.libssh2.dev;
+              zstdSupport = false;
+              idnSupport = false;
+            }).overrideAttrs (oldAttrs:
+              let confFlags = oldAttrs.configureFlags;
+              in {
+                configureFlags = (muslPkgs.lib.take 13 confFlags)
+                  ++ (muslPkgs.lib.drop 14 confFlags)
+                  ++ [ (muslPkgs.lib.withFeature true "libssh2") ];
```
Comment (alxiong) [inline flake.nix]: reminds me why do we need these openssl and libssh for? did you try removing them?

**Review by alxiong — COMMENTED** [review]
Generally LGTM, maybe add instructions in README about developers inside nix.

btw @akonring what command should I run to test? if we are using docker for trying out. if we change any code in this repo, we would reproduce a new docker image? at what time would we need this nix shell for?

**Comment by akonring** [comment]
The nix files here are copy-paste from our zkevm-node (https://github.com/EspressoSystems/zkevm-node/blob/hotshot-integration/flake.nix) which, in turn, are just using the same as the espresso-sequencer but adding a few extra components:

https://github.com/EspressoSystems/cdk-validium-node/blob/8eb070cf2b68fdf26733e6820c6997122a9b2c07/flake.nix#L95-L98

We could:
1. Do without nix completely (My initial thought was that nix is a part of our standard git setup).
2. Put some time in slimming down the nix files to only include the essential parts (I might be missing something but most of it seems superfluous wrt. cdk-node code).
3. Leave it as is.

Let me know what you think, @alxiong.

File: flake.nix (hunk 1 above)
Comment (akonring) [reply flake.nix]: don't think they are used (see below)

**Comment by alxiong** [comment]
i see. my personal preference is to "build up" rather than "trim down". we can simply start with `pre-commit` and the go, docker, postgres. (don't even need any rust toolchain here)

hope it's not too much work, happy to hop on a pair-programming session to work on it.

**Review by alxiong — APPROVED** [review]
Feel free to merge on your command.

thanks for taking the initiative!

we can slowly add what's necessary into `flake.nix`, also current CI didn't really run anything, but we will address that in future PR, since we don't really know well what best to test in CI yet.

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,57 @@
+{
+  description = "cdk-validium-node, with go development environment";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+    flake-compat.url = "github:edolstra/flake-compat";
+    flake-compat.flake = false;
+    pre-commit-hooks.url = "github:cachix/pre-commit-hooks.nix";
```
Comment (akonring) [inline flake.nix]: Should we use input follows?
```suggestion
    pre-commit-hooks = { 
      url = "github:cachix/pre-commit-hooks.nix";
      inputs = {
        nixpkgs.follows = "nixpkgs";
        flake-compat.follows = "flake-compat";
      };
    };
```

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            '';
+          in
+          mkShell {
+            buildInputs = [
+              go
+              protobuf
+              docker
+              docker-compose
+
+              git
```
Comment (akonring) [inline flake.nix:52]: nit: is there a special reason to include this? I'm not sure I've seen it in our other `flake.nix` files

Comment (alxiong) [reply flake.nix:52]: Yah just think it's basic need. I guess system git also works. But it's a small binary.

File: flake.nix (hunk 2 above)
Comment (alxiong) [reply flake.nix]: sure, we can use consistent version of `nixpkgs` and `flake-compat`.

**Pull request merged**